APMrover2/mode_auto.cpp

```diff
@@ -50,13 +44,16 @@ void ModeAuto::update()
                 _reached_destination = true;
             }
             // determine if we should keep navigating
-            if (!_reached_destination || (rover.is_boat() && !near_wp)) {
+            if (!_reached_destination) {
                 // continue driving towards destination
                 calc_steering_to_waypoint(_reached_destination ? rover.current_loc : _origin, _destination, _reversed);
                 calc_throttle(calc_reduced_speed_for_turn_or_distance(_reversed ? -_desired_speed : _desired_speed), true, true);
             } else {
-                // we have reached the destination so stop
-                stop_vehicle();
+                if (rover.is_boat() && !start_loiter()) {
+                    stop_vehicle();
+                } else {
+                    stop_vehicle();
```

Comment:
This logic is weird, If all you're trying to do is set Auto_Loiter then do it here or abstract this whole logic into the start_loiter() function. As-is, stop_vehicle() is called no matter what so this whole thing turns into:
```
if (rover.is_boat()) {
  start_loiter();
}
stop_vehicle();
```

Reply:
@magicrub Thank you for your feedback! I don't want to call stop_vehicle() when start_loiter() returns true.
(rover.is_boat() && !start_loiter()) || stop_vehicle();
Should I not write like this? If so, I will fix it as you pointed out.